pkg/util/workqueue/delaying_queue.go

```diff
@@ -177,7 +182,32 @@ func (q *delayingType) waitingLoop() {
 // inserts the given entry into the sorted entries list
 // same semantics as append()... the given slice may be modified,
 // and the returned value should be used
-func insert(entries []waitFor, entry waitFor) []waitFor {
+func insert(entries []waitFor, knownEntries map[t]time.Time, entry waitFor) []waitFor {
+	// if the entry is already in our retry list and the existing time is before the new one, just skip it
+	existingTime, exists := knownEntries[entry.data]
+	if exists && existingTime.Before(entry.readyAt) {
+		return entries
+	}
+
+	// if the entry exists and is scheduled for later, go ahead and remove the entry
+	if exists {
+		startingIndex := sort.Search(len(entries), func(i int) bool {
+			return entries[i].readyAt.After(existingTime) || existingTime == entries[i].readyAt
```

Comment:
> !entries[i].readyAt.Before(existingTime)?

Positive logic over negative.  This makes a very clear "greater than or equal to".